packages/apollo-federation/src/composition/compose.ts

```diff
@@ -410,13 +414,18 @@ export function addFederationMetadataToSchemaNodes({
     const isValueType = valueTypes.has(typeName);
     const serviceName = isValueType ? null : owningService;
 
-    namedType.federation = {
-      ...namedType.federation,
+    const federationMetadata: FederationType = {
+      ...getFederationMetadata(namedType),
       serviceName,
       isValueType,
       ...(keyDirectivesMap[typeName] && {
         keys: keyDirectivesMap[typeName],
       }),
+    }
+
+    namedType.extensions = {
+      ...namedType.extensions,
+      federation: federationMetadata,
     };
```

Comment:
In the event that the `extensions` property/object already exists (does it always, in newer `graphql` versions?), I can imagine that it might be beneficial to avoid creating new `extensions` objects and instead maintain referential equality to the existing property.  For example, this could be relevant if we're federating existing schemas.

Might we consider merely adding the `federation` property to `extensions`?  If it's indeed `undefined` until an implementor tries writing to it, we could default to creating it in that case?

Reply:
Note: we spoke offline about this and decided we'd move forward with this solution until it becomes a problem.

Sadly we can't mutate the extensions object itself as it's `Readonly`, so instead it must be recreated regardless of whether or not it already exists (if we're respecting the types, anyway).

For posterity, @IvanGoncharov mentioned that we can use `transformSchema` here. This may be the way forward for types, but it leaves us with the same solution for a field's `extensions` below. I'm also hesitant to use `transformSchema` here since it seems to require a rework of the existing approach. Rather than modify the types one at a time, I think we would need to collect a map of types and their new `extensions` during the for loop, followed by a `transformSchema` once we've mapped over the types.